pkg/miniogw/gateway-storj.go

```diff
@@ -5,31 +5,143 @@ package storj
 
 import (
 	"context"
+	"crypto/sha256"
+	"flag"
+	"fmt"
 	"io"
+	"os"
+	"path"
+	"path/filepath"
+	"strconv"
 	"time"
 
 	"github.com/minio/cli"
-
-	"github.com/minio/minio/cmd"
+	minio "github.com/minio/minio/cmd"
 	"github.com/minio/minio/pkg/auth"
 	"github.com/minio/minio/pkg/hash"
+	"github.com/vivint/infectious"
+
+	"storj.io/storj/pkg/eestream"
+)
+
+var (
+	pieceBlockSize = flag.Int("piece_block_size", 4*1024, "block size of pieces")
+	key            = flag.String("key", "a key", "the secret key")
+	rsk            = flag.Int("required", 20, "rs required")
+	rsn            = flag.Int("total", 40, "rs total")
 )
 
 func init() {
-	cmd.RegisterGatewayCommand(cli.Command{
+	minio.RegisterGatewayCommand(cli.Command{
 		Name:            "storj",
 		Usage:           "Storj",
 		Action:          storjGatewayMain,
 		HideHelpCommand: true,
 	})
 }
 
+// getbucketscallback returns the buckets list
+func (s *storjObjects) getBuckets() (buckets []minio.BucketInfo, err error) {
+	buckets = make([]minio.BucketInfo, len(s.storj.bucketlist))
+	for i, bi := range s.storj.bucketlist {
+		buckets[i] = minio.BucketInfo{
+			Name:    bi.bucket.Name,
+			Created: bi.bucket.Created,
+		}
+	}
+	return buckets, nil
+}
+
+// putobjectcallback function handles to add the uploaded file to the bucket's file list structure
+func (s *storjObjects) uploadFile(bucket, object string, filesize int64, metadata map[string]string) (result minio.ListObjectsInfo, err error) {
+	var fl []minio.ObjectInfo
+	for i, v := range s.storj.bucketlist {
+		// bucket string comparision
+		if v.bucket.Name == bucket {
+			/* append the file to the filelist */
+			s.storj.bucketlist[i].filelist.file.Objects = append(
+				s.storj.bucketlist[i].filelist.file.Objects,
+				minio.ObjectInfo{
+					Bucket:      bucket,
+					Name:        object,
+					ModTime:     time.Now(),
+					Size:        filesize,
+					IsDir:       false,
+					ContentType: "application/octet-stream",
+				},
+			)
+			/* populate the filelist */
+			f := make([]minio.ObjectInfo, len(s.storj.bucketlist[i].filelist.file.Objects))
+			for j, fi := range s.storj.bucketlist[i].filelist.file.Objects {
+				f[j] = minio.ObjectInfo{
+					Bucket:      v.bucket.Name,
+					Name:        fi.Name,
+					ModTime:     fi.ModTime,
+					Size:        fi.Size,
+					IsDir:       fi.IsDir,
+					ContentType: fi.ContentType,
+				}
+			}
+			fl = f
+			break
+		}
+	}
+	result = minio.ListObjectsInfo{
+		IsTruncated: false,
+		Objects:     fl,
+	}
+	return result, nil
+}
+
+// listfilescallback returns the files list for a bucket
+func (s *storjObjects) getFiles(bucket string) (result minio.ListObjectsInfo, err error) {
+	var fl []minio.ObjectInfo
+	for i, v := range s.storj.bucketlist {
+		if v.bucket.Name == bucket {
+			/* populate the filelist */
+			f := make([]minio.ObjectInfo, len(s.storj.bucketlist[i].filelist.file.Objects))
+			for j, fi := range s.storj.bucketlist[i].filelist.file.Objects {
```

Comment:
I don't understand why you are recreating the ListObjectInfo here. Isn't `s.storj.bucketlist[i].filelist.file.Objects` exactly what has to be returned?

Reply:
you can code this in one line. The implementation I did is to support the future enhancement. Technically I understand all this can be one line and you don't even need this function. I will keep it there for now till, all the features are supported. No change will be made